### PR TITLE
rename variables nb -> num

### LINKED ISSUE #534: Make all abbreviations: (nb --> num)  and (i --> idx)
**Is your feature request related to a problem? Please describe.**
The code currently uses two different abbreviations for number (`nb` and `num`) and index (`idx` and `i`).

An example of `nb`:
`def training_step(self, batch, batch_nb):`
https://williamfalcon.github.io/pytorch-lightning/LightningModule/RequiredTrainerInterface/#training_step

An example of `num`:
`def log_metrics(self, metrics, step_num):`
https://williamfalcon.github.io/pytorch-lightning/Trainer/Logging/#custom-logger

An example of `idx`:
`def training_step(self, batch, batch_nb, optimizer_idx):`
https://williamfalcon.github.io/pytorch-lightning/LightningModule/RequiredTrainerInterface/#training_step

An example of `i`:
`def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_i, second_order_closure=None):`
https://williamfalcon.github.io/pytorch-lightning/Trainer/hooks/#optimizer_step

**Describe the solution you'd like**
I think switching to using consistent abbreviations for both number and index would improve the design quality of the API.

Although `nb` is currently used more than `num`, PyTorch uses `num` ubiquitously (e.g. `torch.nn.RNN(num_layers=...)`), so using `num` would probably make the API feel more familiar to users, and is the abbreviation I would suggest. I'm less sure which is better between `idx` and `i`, but it looks like there are fewer current uses of `i`, so changing those to use `idx` would probably be easier.

I know one of Lightning's design principles is a backward-compatible API, but there are some options for updating argument names with deprecation warnings that could be considered:
https://stackoverflow.com/questions/49802412/how-to-implement-deprecation-in-python-with-argument-alias

Also if this change is ever going to be made, the sooner probably the better, while Lightning is still in its relatively early stages of development.

**Describe alternatives you've considered**
Alternatively, if this would be too much of a breaking change, perhaps this change could be delayed until when a future breaking change is released.

I'm interested to hear others' opinions on this.

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
PARTIAL Fixes #534.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by Borda** [comment]
@elliotwaite could you have look on this (proposed) changes...

**Comment by williamFalcon** [comment]
@Borda all of these changes need to be backward compatible though

**Comment by Borda** [comment]
for that, I am not sure how to handle multiple methods with the same name just with changed parameter name... adding the past parameter name will crash while passing parameters with values only... :/

**Comment by Borda** [comment]
@williamFalcon it seems that method overloading will make it even worse...
https://www.geeksforgeeks.org/python-method-overloading
I do not think that these parameter name changes are big issue for users, but you say so and you want to keep the names, we shall close this PR/change and conclude that #534 won't be implemented...

**Comment by williamFalcon** [comment]
@Borda 
I think we can:
1. add backward compatibility to trainer arg changes.
2. the arg name changes to training_step, etc shouldn't affect users:

image

File: pl_examples/multi_node_examples/multi_node_ddp2_demo.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def main(hparams):
     # ------------------------
     trainer = Trainer(
         gpus=2,
-        nb_gpu_nodes=2,
+        num_gpu_nodes=2,
```
Comment (williamFalcon) [inline pl_examples/multi_node_examples/multi_node_ddp2_demo.py]: maybe shorten to:
num_nodes?

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 2]
```diff
@@ -234,7 +234,7 @@ def main(hparams):
     trainer = pl.Trainer(
         default_save_path=hparams.save_path,
         gpus=hparams.gpus,
-        max_nb_epochs=hparams.epochs,
+        max_num_epochs=hparams.epochs,
```
Comment (williamFalcon) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: add backward compatibility

**Comment by Borda** [comment]
> I think we can:
> 1. add backward compatibility to trainer arg changes.
> 2. the arg name changes to training_step, etc shouldn't affect users (see example)
> image

but this assumes that the class name is different, so `class Trainer2(...)`?

**Comment by williamFalcon** [comment]
what??? class A is lightningModule. B is the subclass

**Comment by Borda** [comment]
> what??? class A is lightningModule. B is the subclass

I just do not understand what you suggest, adding subclass of `LightningModule`...
you mean to keep both parameters, like you have here?
https://github.com/williamFalcon/pytorch-lightning/blob/89ececb32ba0cfd810737cb90b2285a27332f5d4/pytorch_lightning/trainer/trainer.py#L59

**Comment by williamFalcon** [comment]
changing 
```python
def training_step(self, batch, batch_nb):
```
to 
```python
def training_step(self, batch, batch_idx):
```

does  not affect the  user

**Comment by williamFalcon** [comment]
@Borda rebase?

**Comment by Borda** [comment]
@williamFalcon done :]

**Pull request merged**